webapp/src/Controller/API/JudgehostController.php

```diff
@@ -812,37 +810,47 @@ public function internalErrorAction(Request $request)
 
         if (in_array($disabled['kind'], ['problem', 'language', 'judgehost']) && $judgingId) {
             // Give back judging if we have to.
-            $this->giveBackJudging((int)$judgingId);
+            $hostname = $request->request->get('hostname');
+            /** @var Judgehost $judgehost */
+            $judgehost = $this->em->createQueryBuilder()
+                ->from(Judgehost::class, 'j')
+                ->select('j')
+                ->andWhere('j.hostname = :hostname')
+                ->setParameter(':hostname', $hostname)
+                ->setMaxResults(1)
+                ->getQuery()
+                ->getSingleResult();
+            $this->giveBackJudging((int)$judgingId, $judgehost);
         }
 
         return $error->getErrorid();
     }
 
     /**
-     * Give back the judging with the given judging ID
-     * @param int $judgingId
+     * Give back the unjudged runs from the judging with the given judging ID
+     * @param int       $judgingId
+     * @param Judgehost $judgehost
      */
-    protected function giveBackJudging(int $judgingId)
+    protected function giveBackJudging(int $judgingId, Judgehost $judgehost)
     {
         /** @var Judging $judging */
         $judging = $this->em->getRepository(Judging::class)->find($judgingId);
         if ($judging) {
-            $this->em->transactional(function () use ($judging) {
-                $judging
-                    ->setValid(false)
-                    ->setRejudging(null);
-
-                $judging->getSubmission()->setJudgehost(null);
-
-                // Give back judging, create a new one.
-                $newJudging = new Judging();
-                $newJudging
-                    ->setContest($judging->getContest())
-                    ->setSubmission($judging->getSubmission());
-                $this->em->persist($newJudging);
-                $this->em->flush();
+            $this->em->transactional(function () use ($judging, $judgehost) {
+                /** @var JudgingRun $run */
+                foreach ($judging->getRuns() as $run) {
+                    // We do not have to touch any finished runs
+                    if ($run->getRunresult() !== null) {
+                        continue;
+                    }
 
-                $this->dj->maybeCreateJudgeTasks($newJudging);
+                    // For the other runs, we need to reset the judge task if it belongs to the current judgehost
+                    if ($run->getJudgetask()->getHostname() === $judgehost->getHostname()) {
+                        $run->getJudgetask()->setHostname(null);
+                    }
+                }
+
+                $this->em->flush();
             });
 
             $this->dj->auditlog('judging', $judgingId, 'given back', null,
```

Comment:
should we also update this?

Reply:
What do you propose?

Reply:
Perhaps mention the judge host in the message?